### ISSUE TITLE
Failed test error messages should be printed at the end of the output

### ISSUE THREAD

**Issue by Nick-Mazuk** [oven-sh/bun#2109]
### What is the problem this feature would solve?

When a test fails, I want to be able to identify the failure as quickly as possible to fix it. Currently when a test errors, the error message is printed inline with the output whenever the test runner decides to run the test.

As a result, in a codebase with hundreds of tests (and growing), I can quickly see _when_ a test fails but not _which_ test fails. To find which test that fails, I must scroll upwards.

I've worked in codebases with tens of thousands of tests, which makes this workflow unpracticle.

### What is the feature you are proposing to solve the problem?

By printing all test errors at the bottom of the output, I can quickly view which tests are failing and why they're failing regardless of the number of tests in the repository.

**Comment by simylein** [comment]
Any progress on this? This would be super helpful when working on larger test suites.

**Comment by jakeboone02** [comment]
Copying https://github.com/oven-sh/bun/issues/3342#issuecomment-1602990605 here as it's worth seeing the difference:

> Jest shows the pass/fail result for each test inline, but batches the logging of all the verbose stuff (Expected/Received/etc.) at the end of the output. `bun test` puts the verbose failure explanations interspersed with the one-line pass/fail logs. Comparison of the same test suite as above, with a couple of intentional failures:
>
> ![image](https://github.com/oven-sh/bun/assets/366438/c94f8496-c8d1-4825-aa54-83e7e6fb708b)
>
> With Jest's format it's much easier to see everything that needs to be addressed in one place without  much scrolling, if any.

**Comment by simylein** [comment]
Any updates on this?

**Comment by jakeboone02** [comment]
@simylein This is partially implemented already, but in a weird way. It lists the names of `.skip`'d, `.todo`'d, and failed tests, _but only if more than 20 tests pass_. Relevant code:

https://github.com/oven-sh/bun/blob/c2a5451e93bae70d25012bba3aeac51f0756df63/src/cli/test_command.zig#L926-L958

**Comment by jakeboone02** [comment]
Another thing I just noticed: the changes in #12189 (which I assume will be included in v1.1.19) make the output less useful because the `describe` scope names are not printed in front of the test names. This can probably be addressed at the same time as #12378.

**Comment by simylein** [comment]
> @simylein This is partially implemented already, but in a weird way. It lists the names of any `.skip`'d, `.todo`'d, and failed tests at the end of all the results, but only if more than 20 tests pass. Relevant code:
> 
> 
> 
> https://github.com/oven-sh/bun/blob/c2a5451e93bae70d25012bba3aeac51f0756df63/src/cli/test_command.zig#L926-L958
> 
> 
> 
> So, for example, if you have 20 tests where 15 pass and 5 fail, the names won't be printed because less than 20 passed. But if you have 22 tests and only 1 fails, the failing test will be listed at the end because more than 20 passed.

Yes, I noticed that failed tests get printed last as I am often working in a code base with 400+ unit tests. It's more the scrolling up to find the toEqual diffs which is annoying, since the test output of failed tests itself does not get printed last.

**Comment by danielv-funnel** [comment]
I think this makes sense even for smaller test suites also. Have been adopting bun for a while and think this is the biggest pain for me at least.

I have around 140 tests in current solution and still need to narrow down by filter on a file and getting for example 30 tests and even then it's very annoying to scroll up (I'm using the terminal in vs code which is small part of screen) as even then it cannot fit on the screen. 

You can use `--bail` to stop on first error, but then the `--watch` mode does not work

So I would say this is something jest & vitest has done much better devX on, would be super helpful to have parity on this 🙏 

Not sure if anyone on the bun team sees this for the trees in the forest so adding a ping to @Jarred-Sumner does this make sense to fix?
